File: 199_right_side_view/other.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
  return [] if root.nil?
  result = []
  stack = [[0,root]]
  while !stack.empty?
    height, node = stack.pop
    if result.size == height
      result << node.val
    end
    stack.push [height+1, node.left] if node.left
    stack.push [height+1, node.right] if node.right
  end
  return result
end

File: 199_right_side_view/other.java
package leetcode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Solution299 {

  public List<Integer> rightSideView(TreeNode root) {
    List<Integer> result = new ArrayList<>();
    rightMostAtHeight(root, result, 0)
    return result
  }

  public void rightMostAtHeight(TreeNode node, List<Integer> result, int height) {
    if(null == node) return;
    if (height == result.size) {
      result.add(node.val);
    }
    rightSideView(node.right, result, height+1);
    rightSideView(node.left, result, height+1);
  }

}